blockstack-vm/src/contexts.rs

```diff
@@ -0,0 +1,97 @@
+use std::collections::HashMap;
+use std::collections::HashSet;
+
+use super::types::{DefinedFunction, FunctionIdentifier, Value};
+use super::database::ContractDatabase;
+
+pub struct Environment <'a> {
+    pub global_context: Context <'a>,
+    pub call_stack: CallStack,
+    pub database: Box<ContractDatabase>
+}
+
+impl <'a> Environment <'a> {
+    pub fn new(database: Box<ContractDatabase>) -> Environment<'a> {
+        let global_context = Context::new();
+        Environment {
+            global_context: global_context,
+            call_stack: CallStack::new(),
+            database: database
+        }
+    }
+}
+
+pub struct Context <'a> {
+    pub parent: Option< &'a Context<'a>>,
+    pub variables: HashMap<String, Value>,
+    pub functions: HashMap<String, Box<DefinedFunction>>,
+}
+
+impl <'a> Context <'a> {
+    pub fn new() -> Context<'a> {
+        Context { parent: Option::None,
+                  variables: HashMap::new(),
+                  functions: HashMap::new() }
+    }
+    
+    pub fn extend(&'a self) -> Context<'a> {
```

Comment:
Is there a limit to how many times we can call `extend()`?

Reply:
No -- but if we want to impose a maximum stack depth (which we should), the place to do it at the `apply` function.